fastavro/_writer.pyx

```diff
@@ -449,12 +449,22 @@ cpdef write_union(bytearray fo, datum, schema):
             raise ValueError(msg)
     else:
         pytype = type(datum)
+        best_match_index = -1
```

Comment:
@barrywhart Should `best_match_index`, `most_fields`, and `index` be defined as `cdef int`? I noticed that it ends up with more pure c code when running `cython -a`?

Reply:
I would say yes.  `cython -a` gives a pretty good rough indication of how the code will perform, and more yellow is generally bad.

Reply:
I've changed `best_match_index`, `index`, `most_fields` and `fields` to be declared by `cdef int` as suggested.